src/test/java/com/myhome/services/unit/UserSDJpaServiceTest.java

```diff
@@ -0,0 +1,189 @@
+
+    given(userRepository.findByEmail(request.getEmail()))
+        .willReturn(null);
+    given(passwordEncoder.encode(request.getPassword()))
+        .willReturn(request.getPassword());
+    given(userMapper.userDtoToUser(request))
+        .willReturn(resultUser);
+    given(userRepository.save(resultUser))
+        .willReturn(resultUser);
+    given(userMapper.userToUserDto(resultUser))
+        .willReturn(response);
+
+    // when
+    Optional<UserDto> createdUserDto = userService.createUser(request);
+
+    // then
+    assertTrue(createdUserDto.isPresent());
+    assertEquals(createdUserDto.get(), response);
+    assertEquals(createdUserDto.get().getCommunityIds().size(), 0);
+    verify(userRepository).findByEmail(request.getEmail());
+    verify(passwordEncoder).encode(request.getPassword());
+    verify(userRepository).save(resultUser);
+  }
+
+  @Test
+  void createUserEmailExists() {
+    // given
+    UserDto request = getDefaultUserDtoRequest();
+    User user = new User(request.getName(), request.getUserId(), request.getEmail(), request.getEncryptedPassword());
+
+    given(userRepository.findByEmail(request.getEmail()))
+        .willReturn(user);
+
+    // when
+    Optional<UserDto> createdUserDto = userService.createUser(request);
+
+    // then
+    assertFalse(createdUserDto.isPresent());
+    verify(userRepository).findByEmail(request.getEmail());
+  }
+
+  @Test
+  void getUserDetailsSuccess() {
+    // given
+    UserDto userDto = getDefaultUserDtoRequest();
+    User user = new User(userDto.getName(), userDto.getUserId(), userDto.getEmail(), userDto.getEncryptedPassword());
+
+    given(userRepository.findByUserId(USER_ID))
+        .willReturn(user);
+    given(communityService.listAll())
+        .willReturn(new HashSet<>());
+    given(userMapper.userToUserDto(user))
+        .willReturn(userDto);
+
+    // when
+    Optional<UserDto> createdUserDto = userService.getUserDetails(USER_ID);
+
+    // then
+    assertTrue(createdUserDto.isPresent());
+    assertEquals(createdUserDto.get(), userDto);
+    assertEquals(createdUserDto.get().getCommunityIds().size(), 0);
+    verify(userRepository).findByUserId(USER_ID);
+    verify(communityService).listAll();
+  }
+
+  @Test
+  void getUserDetailsSuccessWithCommunityIds() {
+    // given
+    UserDto userDto = getDefaultUserDtoRequest();
+    User user = new User(userDto.getName(), userDto.getUserId(), userDto.getEmail(), userDto.getEncryptedPassword());
+
+    CommunityAdmin communityUserAdmin = getAdminFromUser();
+    Community firstCommunity = createCommunityWithUserAdmin(communityUserAdmin);
+    Community secCommunity = createCommunityWithUserAdmin(communityUserAdmin);
+
+    Set<Community> communities = new HashSet<Community>(){{
+      add(firstCommunity);
+      add(secCommunity);
+    }};
+    Set<String> communitiesIds = communities.stream().map(comm -> comm.getCommunityId()).collect(Collectors.toSet());
```

Comment:
Code can be reformatted to better readability and method reference can be used (to exchange `comm.getCommunityId()`).

```
Set<String> communitiesIds = communities.stream()
        .map(Community::getCommunityId)
        .collect(Collectors.toSet());
```